src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -3137,5 +3132,346 @@ static void M()
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "Unused").WithArguments("Program.Unused").WithLocation(4, 24)
+                );
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void AttributesMarkUsed()
+        {
+            var source = @"
+class A : System.Attribute
+{
+    public A(string x) { }
+}
+class Program
+{
+    static void Main()
+    {
+        int x;
+        void AttrParam([A(nameof(x))] int a) { }
+        void AttrParamLocfunc([A(nameof(AttrParam))] int a) { }
+        AttrParamLocfunc(2);
+        int y;
+        void AttrType<[A(nameof(y))] T>(int a) { }
+        void AttrTypeLocfunc<[A(nameof(AttrType))] T>(int a) { }
+        AttrTypeLocfunc<int>(2);
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: DefaultParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (11,24): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrParam([A(nameof(x))] int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(x))]").WithLocation(11, 24),
+                // (12,31): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrParamLocfunc([A(nameof(AttrParam))] int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(AttrParam))]").WithLocation(12, 31),
+                // (15,23): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrType<[A(nameof(y))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(y))]").WithLocation(15, 23),
+                // (15,33): error CS0103: The name 'y' does not exist in the current context
+                //         void AttrType<[A(nameof(y))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "y").WithArguments("y").WithLocation(15, 33),
+                // (16,30): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrTypeLocfunc<[A(nameof(AttrType))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(AttrType))]").WithLocation(16, 30),
+                // (16,40): error CS0103: The name 'AttrType' does not exist in the current context
+                //         void AttrTypeLocfunc<[A(nameof(AttrType))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "AttrType").WithArguments("AttrType").WithLocation(16, 40),
+                // (10,13): warning CS0168: The variable 'x' is declared but never used
+                //         int x;
+                Diagnostic(ErrorCode.WRN_UnreferencedVar, "x").WithArguments("x").WithLocation(10, 13),
```

Comment:
> Let me know if that's not acceptable. 

I think this is not acceptable. For the purpose of the diagnostics that we are fixing, it shouldn't matter whether there are other errors in the code.